polynote-spark/src/main/scala/polynote/kernel/remote/protocol.scala

```diff
@@ -0,0 +1,156 @@
+}
+
+final case class QueueCell(reqId: Int, id: CellID) extends RemoteRequest
+object QueueCell extends RemoteRequestCompanion[QueueCell](3)
+
+final case class CompletionsAt(reqId: Int, id: CellID, pos: Int) extends RemoteRequest
+object CompletionsAt extends RemoteRequestCompanion[CompletionsAt](5)
+
+final case class ParametersAt(reqId: Int, id: CellID, pos: Int) extends RemoteRequest
+object ParametersAt extends RemoteRequestCompanion[ParametersAt](6)
+
+final case class CurrentSymbols(reqId: Int) extends RemoteRequest
+object CurrentSymbols extends RemoteRequestCompanion[CurrentSymbols](7)
+
+final case class CurrentTasks(reqId: Int) extends RemoteRequest
+object CurrentTasks extends RemoteRequestCompanion[CurrentTasks](8)
+
+final case class IdleRequest(reqId: Int) extends RemoteRequest
+object IdleRequest extends RemoteRequestCompanion[IdleRequest](9)
+
+final case class InfoRequest(reqId: Int) extends RemoteRequest
+object InfoRequest extends RemoteRequestCompanion[InfoRequest](10)
+
+final case class GetHandleDataRequest(reqId: Int, handleType: HandleType, handle: Int, count: Int) extends RemoteRequest
+object GetHandleDataRequest extends RemoteRequestCompanion[GetHandleDataRequest](11)
+
+final case class ModifyStreamRequest(reqId: Int, handleId: Int, ops: List[TableOp]) extends RemoteRequest
+object ModifyStreamRequest extends RemoteRequestCompanion[ModifyStreamRequest](12) {
+  import TableOpCodec.tableOpCodec
+  implicit val codec: Codec[ModifyStreamRequest] = cachedImplicit
+}
+
+final case class ReleaseHandleRequest(reqId: Int, handleType: HandleType, handleId: Int) extends RemoteRequest
+object ReleaseHandleRequest extends RemoteRequestCompanion[ReleaseHandleRequest](13)
+
+final case class CancelTasksRequest(reqId: Int) extends RemoteRequest
+object CancelTasksRequest extends RemoteRequestCompanion[CancelTasksRequest](14)
+
+final case class UpdateNotebookRequest(reqId: Int, version: Int, update: NotebookUpdate) extends RemoteRequest
+object UpdateNotebookRequest extends RemoteRequestCompanion[UpdateNotebookRequest](15) {
+  private implicit val notebookUpdateCodec: Codec[NotebookUpdate] = Message.codec.exmap[NotebookUpdate](
+    _ match {
+      case msg: NotebookUpdate => Attempt.successful(msg)
+      case _ => Attempt.failure(Err("Not a notebook update message"))
+    },
+    msg => Attempt.successful(msg)
+  )
+
+  implicit val codec: Codec[UpdateNotebookRequest] = cachedImplicit
+}
+
+object RemoteRequest {
+  implicit val discriminated: Discriminated[RemoteRequest, Byte] = Discriminated(byte)
+  implicit val codec: Codec[RemoteRequest] = cachedImplicit
+}
+
+
+sealed trait RemoteResponse
+sealed trait RequestResponse extends RemoteResponse {
+  def reqId: Int
+}
+
+abstract class RemoteResponseCompanion[T <: RemoteResponse](msgTypeId: Byte) {
+  implicit val discriminator: Discriminator[RemoteResponse, T, Byte] = Discriminator(msgTypeId)
+}
+
+final case class UnitResponse(reqId: Int) extends RequestResponse
```

Comment:
So this is essentially an ACK?

Reply:
Essentially... the server needs to know that the `IO[Unit]` completed.